includes/classes/shopping_cart.php

```diff
@@ -1959,15 +1959,15 @@ function actionAddProduct($goto, $parameters) {
         } else {
           // process normally
           // iii 030813 added: File uploading: save uploaded files with unique file names
-          $real_ids = isset($_POST['id']) ? $_POST['id'] : "";
+          $real_ids = isset($_POST['id']) ? $_POST['id'] : array();
           if (isset($_GET['number_of_uploads']) && $_GET['number_of_uploads'] > 0) {
             /**
              * Need the upload class for attribute type that allows user uploads.
              *
              */
             include(DIR_WS_CLASSES . 'upload.php');
             for ($i = 1, $n = $_GET['number_of_uploads']; $i <= $n; $i++) {
-              if (zen_not_null($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]]) and ($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]] != 'none')) {
+              if (isset($_POST[UPLOAD_PREFIX . $i]) && isset($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]]) && zen_not_null($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]]) && (!isset($_POST[UPLOAD_PREFIX . $i]) || !isset($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]]) || ($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]] != 'none'))) {
```

Comment:
Can `isset($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]]) && zen_not_null($_FILES['id']['tmp_name'][TEXT_PREFIX . $_POST[UPLOAD_PREFIX . $i]])` 
be shortened with just `!empty` instead of `isset && zen_not_null`?

Reply:
If the simple assumption is that all files will have some sort of non-empty extension even for a file named 0, then yes could substitute with `!empty`.  At the time of writing I had briefly considered it as done in other similar rewrites, but opted to leave as seen, awaiting further discussion, potential testing for that case and evaluation.

By the way, at the time of writing, I was unaware of a plugin that performs a similar function with additional "notifications".  I have not fully reviewed that code, but would not think that it should be discounted potentially as a more complete solution.